README.md
[![js-semistandard-style](https://img.shields.io/badge/code%20style-semistandard-brightgreen.svg)](https://github.com/standard/semistandard)
[![Conventional Commits](https://img.shields.io/badge/Conventional%20Commits-1.0.0-yellow.svg)](https://conventionalcommits.org)

# About

A unified way to emit responses for [1C:Enterprise](https://www.1ci.com/developers/) platform.

Allows to generate both **sync** and **async** responses.

# Installation

```bash
npm install @skripio/response-emitter
```

# Usage

1.  Import `ResponseEmitter` class and instantiate emitter object. Emitter will create `div` element with id provided and attach it to `body` by default. That can be changed by passing `parentElementSelector` to the constructor.

```JavaScript
    import ResponseEmitter from '@skripio/response-emitter';
    const re = new ResponseEmitter({
        responseElementId: 'MyComponentName', 
        responseElementClass: 'response'
        });
```

1.  Call object methods to emit response message or to do auxiliary stuff.

# Response

**Skripio** `ResponseEmitter` prepares and emits response message object to [1C:Enterprise](https://www.1ci.com/developers/) platform.

Response message is guaranteed to have following structure:

```json
    { 
        "callback": "callback identifier", 
        "code": "response code", 
        "payload": "message payload" 
    }
```

where:

*   **callback** - contains callback identifier that allows to link function calls to responses when executing asynchronous code in `JS` from `1C:Enterprise`.
*   **code** - Response code.
*   **payload** - Message payload.

# Reference

<!-- Generated by documentation.js. Update this documentation by updating the source code. -->

### Table of Contents

*   [ResponseEmitter](#responseemitter)
    *   [Parameters](#parameters)
    *   [emitResponse](#emitresponse)
    *   [codes](#codes)
    *   [getUserMessage](#getusermessage)

## ResponseEmitter

**The response emitter object.**

### Parameters

*   `args` **[object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)** Constructor arguments object.

    *   `args.responseElementId` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Id that will be assigned to the response DOM element.
    *   `args.responseElementClass` **([string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String) | [Array](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array)<[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)>)** CSS class(es) that will be assigned to the response DOM element.
    *   `args.parentElementSelector` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Selector of a DOM element that response DOM element will be attached to. (optional, default `'body'`)

### emitResponse

**Generates serialized response message object. Emits `click` event if required.**

#### Parameters

*   `callback` **any** Any attribute to be placed against id key of the message object.
*   `code` **any** Any attribute to be placed against code key of the message object.
*   `payload` **any** Any attribute to be placed against payload key of the message object.
*   `click` **[boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)** If truthy then click event with serialized message object will be emitted. No event emitting otherwise. (optional, default `false`)

Returns **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Response message object serialized.

### codes

**`ResponseEmitter.codes` constants.**

*   **DONE**        - 200
*   **RESULT**      - 201
*   **USER_ERROR**  - 300
*   **DEV_ERROR**   - 400

### getUserMessage

**`ResponseEmitter.getUserMessage` static method generates 1C formatted notification message from phrases in different languages.**

#### Parameters

*   `phrases` **[object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)** Phrases Object that contains phrases in different languages where **key** must be language code and **value** contains phrase in that language.

#### Examples

```javascript
{
  ru: 'Сообщение на Русском языке.',
  en: 'Message in English language.'
}
```

Returns **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Formatted text or an empty string if object provided contains no keys with language codes listed in [ISO 639-1](https://en.wikipedia.org/wiki/List_of_ISO\_639-1\_codes).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.0.5](https://github.com/fmecgroup/skripio-helper-response-emitter/compare/v1.0.4...v1.0.5) (2021-10-21)


### Bug Fixes

* Rename id into callback ([55b3374](https://github.com/fmecgroup/skripio-helper-response-emitter/commit/55b3374c72f4e53bb4b5962b9097c07274ff405e))

### [1.0.4](https://github.com/fmecgroup/skripio-helper-response-emitter/compare/v1.0.3...v1.0.4) (2021-09-23)

### [1.0.3](https://github.com/fmecgroup/skripio-helper-response-emitter/compare/v1.0.2...v1.0.3) (2021-09-23)


### Bug Fixes

* Make getUserMessage static. Update string bug. ([395315e](https://github.com/fmecgroup/skripio-helper-response-emitter/commit/395315eb9d9cb7565bdf17f59abfbcd2e777670d)), closes [AB#1953](https://dev.azure.com/fmec/FMS/_workitems/edit/1953) [AB#1954](https://dev.azure.com/fmec/FMS/_workitems/edit/1954)

### [1.0.2](https://github.com/fmecgroup/skripio-helper-response-emitter/compare/v1.0.1...v1.0.2) (2021-09-19)


### Bug Fixes

* Create and attach response DOM element. ([32789e8](https://github.com/fmecgroup/skripio-helper-response-emitter/commit/32789e8888045f449985c5b72912e5cde317f42f)), closes [AB#1934](https://dev.azure.com/fmec/FMS/_workitems/edit/1934)
* Links to AB workitems in changelog. ([5141285](https://github.com/fmecgroup/skripio-helper-response-emitter/commit/5141285eca6b25a000422506ec93222af4686196)), closes [AB#1934](https://dev.azure.com/fmec/FMS/_workitems/edit/1934)

### [1.0.1](https://github.com/fmecgroup/skripio-helper-response-emitter/compare/v1.0.0...v1.0.1) (2021-09-18)


### Bug Fixes

* Rename `message` key to `payload` ([5b00bdb](https://github.com/fmecgroup/skripio-helper-response-emitter/commit/5b00bdbd9ff9f01e238a3bb3bb384b2bb28c90a6)), closes [AB#1927](https://dev.azure.com/fmec/FMS/_workitems/edit//1927)

## 1.0.0 (2021-09-15)


### Features

* add class and it's methods ([fa6031e](https://github.com/fmecgroup/skripio-helper-response-emitter/commit/fa6031e914fbda99c77e566f23cde957bc60b9db)), closes [AB#1731](https://dev.azure.com/fmec/FMS/_workitems/edit//1731)


modules/languages.js
'use strict';
export default [
  { code: 'ab', name: 'Abkhazian' },
  { code: 'aa', name: 'Afar' },
  { code: 'af', name: 'Afrikaans' },
  { code: 'ak', name: 'Akan' },
  { code: 'sq', name: 'Albanian' },
  { code: 'am', name: 'Amharic' },
  { code: 'ar', name: 'Arabic' },
  { code: 'an', name: 'Aragonese' },
  { code: 'hy', name: 'Armenian' },
  { code: 'as', name: 'Assamese' },
  { code: 'av', name: 'Avaric' },
  { code: 'ae', name: 'Avestan' },
  { code: 'ay', name: 'Aymara' },
  { code: 'az', name: 'Azerbaijani' },
  { code: 'bm', name: 'Bambara' },
  { code: 'ba', name: 'Bashkir' },
  { code: 'eu', name: 'Basque' },
  { code: 'be', name: 'Belarusian' },
  { code: 'bn', name: 'Bengali' },
  { code: 'bh', name: 'Bihari languages' },
  { code: 'bi', name: 'Bislama' },
  { code: 'bs', name: 'Bosnian' },
  { code: 'br', name: 'Breton' },
  { code: 'bg', name: 'Bulgarian' },
  { code: 'my', name: 'Burmese' },
  { code: 'ca', name: 'Catalan, Valencian' },
  { code: 'km', name: 'Central Khmer' },
  { code: 'ch', name: 'Chamorro' },
  { code: 'ce', name: 'Chechen' },
  { code: 'ny', name: 'Chichewa, Chewa, Nyanja' },
  { code: 'zh', name: 'Chinese' },
  { code: 'cu', name: 'Church Slavonic, Old Bulgarian, Old Church Slavonic' },
  { code: 'cv', name: 'Chuvash' },
  { code: 'kw', name: 'Cornish' },
  { code: 'co', name: 'Corsican' },
  { code: 'cr', name: 'Cree' },
  { code: 'hr', name: 'Croatian' },
  { code: 'cs', name: 'Czech' },
  { code: 'da', name: 'Danish' },
  { code: 'dv', name: 'Divehi, Dhivehi, Maldivian' },
  { code: 'nl', name: 'Dutch, Flemish' },
  { code: 'dz', name: 'Dzongkha' },
  { code: 'en', name: 'English' },
  { code: 'eo', name: 'Esperanto' },
  { code: 'et', name: 'Estonian' },
  { code: 'ee', name: 'Ewe' },
  { code: 'fo', name: 'Faroese' },
  { code: 'fj', name: 'Fijian' },
  { code: 'fi', name: 'Finnish' },
  { code: 'fr', name: 'French' },
  { code: 'ff', name: 'Fulah' },
  { code: 'gd', name: 'Gaelic, Scottish Gaelic' },
  { code: 'gl', name: 'Galician' },
  { code: 'lg', name: 'Ganda' },
  { code: 'ka', name: 'Georgian' },
  { code: 'de', name: 'German' },
  { code: 'ki', name: 'Gikuyu, Kikuyu' },
  { code: 'el', name: 'Greek (Modern)' },
  { code: 'kl', name: 'Greenlandic, Kalaallisut' },
  { code: 'gn', name: 'Guarani' },
  { code: 'gu', name: 'Gujarati' },
  { code: 'ht', name: 'Haitian, Haitian Creole' },
  { code: 'ha', name: 'Hausa' },
  { code: 'he', name: 'Hebrew' },
  { code: 'hz', name: 'Herero' },
  { code: 'hi', name: 'Hindi' },
  { code: 'ho', name: 'Hiri Motu' },
  { code: 'hu', name: 'Hungarian' },
  { code: 'is', name: 'Icelandic' },
  { code: 'io', name: 'Ido' },
  { code: 'ig', name: 'Igbo' },
  { code: 'id', name: 'Indonesian' },
  { code: 'ia', name: 'Interlingua (International Auxiliary Language Association)' },
  { code: 'ie', name: 'Interlingue' },
  { code: 'iu', name: 'Inuktitut' },
  { code: 'ik', name: 'Inupiaq' },
  { code: 'ga', name: 'Irish' },
  { code: 'it', name: 'Italian' },
  { code: 'ja', name: 'Japanese' },
  { code: 'jv', name: 'Javanese' },
  { code: 'kn', name: 'Kannada' },
  { code: 'kr', name: 'Kanuri' },
  { code: 'ks', name: 'Kashmiri' },
  { code: 'kk', name: 'Kazakh' },
  { code: 'rw', name: 'Kinyarwanda' },
  { code: 'kv', name: 'Komi' },
  { code: 'kg', name: 'Kongo' },
  { code: 'ko', name: 'Korean' },
  { code: 'kj', name: 'Kwanyama, Kuanyama' },
  { code: 'ku', name: 'Kurdish' },
  { code: 'ky', name: 'Kyrgyz' },
  { code: 'lo', name: 'Lao' },
  { code: 'la', name: 'Latin' },
  { code: 'lv', name: 'Latvian' },
  { code: 'lb', name: 'Letzeburgesch, Luxembourgish' },
  { code: 'li', name: 'Limburgish, Limburgan, Limburger' },
  { code: 'ln', name: 'Lingala' },
  { code: 'lt', name: 'Lithuanian' },
  { code: 'lu', name: 'Luba-Katanga' },
  { code: 'mk', name: 'Macedonian' },
  { code: 'mg', name: 'Malagasy' },
  { code: 'ms', name: 'Malay' },
  { code: 'ml', name: 'Malayalam' },
  { code: 'mt', name: 'Maltese' },
  { code: 'gv', name: 'Manx' },
  { code: 'mi', name: 'Maori' },
  { code: 'mr', name: 'Marathi' },
  { code: 'mh', name: 'Marshallese' },
  { code: 'ro', name: 'Moldovan, Moldavian, Romanian' },
  { code: 'mn', name: 'Mongolian' },
  { code: 'na', name: 'Nauru' },
  { code: 'nv', name: 'Navajo, Navaho' },
  { code: 'nd', name: 'Northern Ndebele' },
  { code: 'ng', name: 'Ndonga' },
  { code: 'ne', name: 'Nepali' },
  { code: 'se', name: 'Northern Sami' },
  { code: 'no', name: 'Norwegian' },
  { code: 'nb', name: 'Norwegian Bokmål' },
  { code: 'nn', name: 'Norwegian Nynorsk' },
  { code: 'ii', name: 'Nuosu, Sichuan Yi' },
  { code: 'oc', name: 'Occitan (post 1500)' },
  { code: 'oj', name: 'Ojibwa' },
  { code: 'or', name: 'Oriya' },
  { code: 'om', name: 'Oromo' },
  { code: 'os', name: 'Ossetian, Ossetic' },
  { code: 'pi', name: 'Pali' },
  { code: 'pa', name: 'Panjabi, Punjabi' },
  { code: 'ps', name: 'Pashto, Pushto' },
  { code: 'fa', name: 'Persian' },
  { code: 'pl', name: 'Polish' },
  { code: 'pt', name: 'Portuguese' },
  { code: 'qu', name: 'Quechua' },
  { code: 'rm', name: 'Romansh' },
  { code: 'rn', name: 'Rundi' },
  { code: 'ru', name: 'Russian' },
  { code: 'sm', name: 'Samoan' },
  { code: 'sg', name: 'Sango' },
  { code: 'sa', name: 'Sanskrit' },
  { code: 'sc', name: 'Sardinian' },
  { code: 'sr', name: 'Serbian' },
  { code: 'sn', name: 'Shona' },
  { code: 'sd', name: 'Sindhi' },
  { code: 'si', name: 'Sinhala, Sinhalese' },
  { code: 'sk', name: 'Slovak' },
  { code: 'sl', name: 'Slovenian' },
  { code: 'so', name: 'Somali' },
  { code: 'st', name: 'Sotho, Southern' },
  { code: 'nr', name: 'South Ndebele' },
  { code: 'es', name: 'Spanish, Castilian' },
  { code: 'su', name: 'Sundanese' },
  { code: 'sw', name: 'Swahili' },
  { code: 'ss', name: 'Swati' },
  { code: 'sv', name: 'Swedish' },
  { code: 'tl', name: 'Tagalog' },
  { code: 'ty', name: 'Tahitian' },
  { code: 'tg', name: 'Tajik' },
  { code: 'ta', name: 'Tamil' },
  { code: 'tt', name: 'Tatar' },
  { code: 'te', name: 'Telugu' },
  { code: 'th', name: 'Thai' },
  { code: 'bo', name: 'Tibetan' },
  { code: 'ti', name: 'Tigrinya' },
  { code: 'to', name: 'Tonga (Tonga Islands)' },
  { code: 'ts', name: 'Tsonga' },
  { code: 'tn', name: 'Tswana' },
  { code: 'tr', name: 'Turkish' },
  { code: 'tk', name: 'Turkmen' },
  { code: 'tw', name: 'Twi' },
  { code: 'ug', name: 'Uighur, Uyghur' },
  { code: 'uk', name: 'Ukrainian' },
  { code: 'ur', name: 'Urdu' },
  { code: 'uz', name: 'Uzbek' },
  { code: 've', name: 'Venda' },
  { code: 'vi', name: 'Vietnamese' },
  { code: 'vo', name: 'Volap_k' },
  { code: 'wa', name: 'Walloon' },
  { code: 'cy', name: 'Welsh' },
  { code: 'fy', name: 'Western Frisian' },
  { code: 'wo', name: 'Wolof' },
  { code: 'xh', name: 'Xhosa' },
  { code: 'yi', name: 'Yiddish' },
  { code: 'yo', name: 'Yoruba' },
  { code: 'za', name: 'Zhuang, Chuang' },
  { code: 'zu', name: 'Zulu' }
];


index.js
'use strict';
import languages from './modules/languages.js';

const languageCodes = languages.map((element) => {
  return element.code;
});

/**
* **The response emitter object.**
* @param {object}             args                                    - Constructor arguments object.
* @param {string}             args.responseElementId                  - Id that will be assigned to the response DOM element.
* @param {(string|string[])}  args.responseElementClass               - CSS class(es) that will be assigned to the response DOM element.
* @param {string}             [args.parentElementSelector = 'body']   - Selector of a DOM element that response DOM element will be attached to.
*/
export default class ResponseEmitter {
  constructor ({ responseElementId, responseElementClass, parentElementSelector = 'body' }) {
    let parentDiv;

    try {
      parentDiv = document.querySelector(parentElementSelector);
    } catch (error) {
      throw new Error(error.message);
    }
    if (!parentDiv) {
      throw new Error(`Failed to instantiate ResponseEmitter. '${parentElementSelector}' is not a valid parent element selector.`);
    }
    if (!responseElementId) {
      throw new Error(`Failed to instantiate ResponseEmitter. '${responseElementId}' is not a valid id value.`);
    }

    this._responseDiv = document.createElement('div');
    this._responseDiv.id = responseElementId;

    if (responseElementClass) {
      this._responseDiv.classList.add(responseElementClass);
    }

    parentDiv.appendChild(this._responseDiv);
  }

  /**
  * **`ResponseEmitter.codes` constants.**
  * - **DONE**        - 200
  * - **RESULT**      - 201
  * - **USER_ERROR**  - 300
  * - **DEV_ERROR**   - 400
  */
  static get codes () {
    return {
      DONE: '200',
      RESULT: '201',
      USER_ERROR: '300',
      DEV_ERROR: '400'
    };
  }

  /**
  * **`ResponseEmitter.getUserMessage` static method generates 1C formatted notification message from phrases in different languages.**
  * @param    {object} phrases - Phrases Object that contains phrases in different languages where **key** must be language code and **value** contains phrase in that language.
  * @returns  {string} - Formatted text or an empty string if object provided contains no keys with language codes listed in [ISO 639-1](https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes).
  * @example
  * {
  *   ru: 'Сообщение на Русском языке.',
  *   en: 'Message in English language.'
  * }
  */
  static getUserMessage (phrases) {
    if (!phrases || typeof phrases !== 'object') {
      throw new Error(`'${phrases}' phrase object is invalid.`);
    }

    let result;

    for (const [key, value] of Object.entries(phrases)) {
      if (languageCodes.includes(key)) {
        if (!result) {
          result = `${key} = '${value}'`;
          continue;
        }
        result += `; ${key} = '${value}'`;
      }
    }

    return result;
  }

  /**
  * **Generates serialized response message object. Emits `click` event if required.**
  * @param    {*} callback     - Any attribute to be placed against id key of the message object.
  * @param    {*} code         - Any attribute to be placed against code key of the message object.
  * @param    {*} payload      - Any attribute to be placed against payload key of the message object.
  * @param    {boolean} click  - If truthy then click event with serialized message object will be emitted. No event emitting otherwise.
  * @returns  {string}         - Response message object serialized.
  */
  emitResponse (callback, code, payload, click = false) {
    const result = JSON.stringify({
      callback: callback,
      code: code,
      payload: payload
    });

    if (click) {
      this._responseDiv.innerText = result;
      this._responseDiv.click();
    }

    return result;
  }
}
